### PR TITLE
Refactor MCP composition

### LINKED ISSUE #131: Lifespans that require parameters are difficult
When composing MCP Servers, if your child MCP Servers require initialization of clients it's a bit painful to initialize them without setting up some sort of global state or moving the creation of the mcp server into a function -- which then means you cannot use the tool/resource decorators.

It would be nice if I could do some sort of dependency injection from my parent mcp server to the children.

Either via #55 or by providing extra parameters when mounting the child to the parent mcp server

### LINKED ISSUE #139: Nested MCP Servers still have lifespan issues
After https://github.com/jlowin/fastmcp/issues/127 I've been struggling to get through having a parent MCP server which does CLI/env cli parsing, and creates two child mcp servers passing them settings objects so they don't have to do their own CLI/env parsing.

It looks like the nested MCP servers are being handed the request context of their parent instead of their own lifespan. This means that for child servers to handle requests their parent context needs to know about their context.

### LINKED ISSUE #167: FastMCP add umount method which can remove another app dynamically

### PR DESCRIPTION
**Pull request by jlowin**
This is a significant refactor that contains some breaking changes for FastMCP methods (though not the major user-facing ones). 

FastMCP.mount() now creates a live link between the mounting server and the mounted server. This keeps the mounted server intact, meaning that server-side features like lifespans should be respected, as well as adding tools to the moutned server after it has been mounted. Servers can be unmounted as well.

FastMCP.import_server() is a *one time* operation that copies resources/templates/prompts/tools from the imported server to the main one. This has latency and complexity benefits, but the tradeoff is that later changes to the imported server are not reflected in the importing server, and all server-side functionality like lifespans are lost.

Closes #167 
Closes #139 
Closes #131

### CODE REVIEW

**Comment by jlowin** [comment]
@strawgate, would really appreciate any thoughts you have here since you've probably pushed composition harder than anyone!

**Comment by strawgate** [comment]
Ill convert my mcp server back to using composition tomorrow and will take a look!

File: src/fastmcp/server/openapi.py [hunk 1]
```diff
@@ -613,25 +614,18 @@ def _create_openapi_template(self, route: openapi.HTTPRoute, operation_id: str):
...
-                # Convert dict to User object
-                if isinstance(result, dict):
-                    return User(**result)
-            except ImportError:
-                # If User class not found, just return the raw result
-                pass
+        # For other tools, ensure the response is wrapped in TextContent
+        if isinstance(result, dict | str):
+            if isinstance(result, dict):
+                result_text = json.dumps(result)
```
[the remark this answers is not in the document]
Comment (QuantumAlchemist03) [reply src/fastmcp/server/openapi.py:626]: ✅ Good catch — using json.dumps without importing the module would raise a NameError.

🛠️ Suggested fix: Add import json at the top of the file if not already present.

File: src/fastmcp/prompts/prompt.py [hunk 2]
```diff
@@ -115,7 +115,7 @@ def from_function(
 
         return cls(
             name=func_name,
-            description=description or fn.__doc__ or "",
+            description=description or fn.__doc__,
```
[the remark this answers is not in the document]
Comment (QuantumAlchemist03) [reply src/fastmcp/prompts/prompt.py:118]: 🔴 This change introduces a potential bug where consumers may receive None instead of a string.

🛠️ Recommended change: description = description or fn.__doc__ or ""

**Comment by QuantumAlchemist03** [comment]
✅ Good effort on improving the code and handling JSON formatting.

⚠️ Avoid regressing default handling for potentially None values like docstrings.

💡 Suggest adding automated tests to catch regressions when a None value is passed where a string is expected.

**Comment by jlowin** [comment]
I'm going to merge this because it touches enough files that it's going to cause merge conflicts elsewhere and appears to pass all basic tests but will welcome feedback on ergonomics and issues.

**Comment by strawgate** [comment]
@jlowin I set this up in my mcp server and everything seems to be working perfectly on the very first try 🚀🚀

**Comment by jlowin** [comment]
Yes!! Glad to hear it.
